hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/UpgradeSubcommand.java

```diff
@@ -0,0 +1,194 @@
+    versionProvider = HddsVersionProvider.class)
+public class UpgradeSubcommand implements Callable<Void> {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(UpgradeSubcommand.class);
+
+  @CommandLine.Spec
+  private static CommandLine.Model.CommandSpec spec;
+
+  @CommandLine.Option(names = {"--volume"},
+      required = false,
+      description = "volume name")
+  private String volume;
+
+  @CommandLine.Option(names = {"-y", "--yes"},
+      description = "Continue without interactive user confirmation")
+  private boolean yes;
+
+  private static OzoneConfiguration ozoneConfiguration;
+
+
+  @Override
+  public Void call() throws Exception {
+    final UpgradeChecker upgradeChecker = new UpgradeChecker();
+
+    OzoneConfiguration configuration = getConfiguration();
+
+    Pair<Boolean, String> pair =
+        upgradeChecker.checkDatanodeNotStarted();
+    final boolean dnNotStarted = pair.getKey();
+    if (!dnNotStarted) {
+      out().println(pair.getValue());
+      return null;
+    }
+
+    final Pair<Integer, Integer> layoutVersion =
+        upgradeChecker.getLayoutVersion(configuration);
+    final Integer softwareLayoutVersion = layoutVersion.getLeft();
+    final Integer metadataLayoutVersion = layoutVersion.getRight();
+    final int needLayoutVersion =
+        HDDSLayoutFeature.DATANODE_SCHEMA_V3.layoutVersion();
+
+    if (metadataLayoutVersion < needLayoutVersion ||
+        softwareLayoutVersion < needLayoutVersion) {
+      out().println(String.format(
+          "Please upgrade your software version, no less than %s," +
+              " current metadata layout version is %d," +
+              " software layout version is %d",
+          HDDSLayoutFeature.DATANODE_SCHEMA_V3.name(),
+          metadataLayoutVersion, softwareLayoutVersion));
+      return null;
+    }
+
+    if (!Strings.isNullOrEmpty(volume)) {
+      configuration.set(ScmConfigKeys.HDDS_DATANODE_DIR_KEY, volume);
+    }
+
+    DatanodeDetails dnDetail =
+        UpgradeUtils.getDatanodeDetails(configuration);
+
+    final HddsProtos.NodeOperationalState opState =
+        dnDetail.getPersistedOpState();
+
+    if (!opState.equals(HddsProtos.NodeOperationalState.IN_MAINTENANCE)) {
+      out().println("This command requires the datanode's " +
+          "NodeOperationalState to be IN_MAINTENANCE, currently is " +
+          opState);
+      return null;
+    }
+
+    List<HddsVolume> allVolume =
+        upgradeChecker.getAllVolume(dnDetail, configuration);
+
+    final List<File> volumeDBPath = upgradeChecker.getVolumeDBPath(allVolume);
+
+    if (volumeDBPath.isEmpty()) {
+      out().println("No volume db store exists, need finalizae data node.");
+      return null;
+    }
+
+    for (HddsVolume hddsVolume : allVolume) {
+      if (UpgradeChecker.checkAlreadyMigrate(hddsVolume)) {
+        out().println("Volume " + hddsVolume.getVolumeRootDir() +
+            " it's already upgraded, skip it.");
+        return null;
```

Comment:
Shall we continue here, instead of return null?

Reply:
I think we should put the upgrade complete file check,  upgrade lock file check, and upgrade lock file creation together. Otherwise, there will be chance that when in process A,  upgrade complete file is checked, it's not there, but there is another process B , which is doing the UpgradeTask at that time, and process B finishes the UpgradeTask before process A begins the UpgradeTask task, then process A will do the same upgrade work one more time.

Reply:
sure.

> I think we should put the upgrade complete file check, upgrade lock file check, and upgrade lock file creation together. Otherwise, there will be chance that when in process A, upgrade complete file is checked, it's not there, but there is another process B , which is doing the UpgradeTask at that time, and process B finishes the UpgradeTask before process A begins the UpgradeTask task, then process A will do the same upgrade work one more time.